sessions.ansible-test-units: Missing Collection Dependencies

## Summary

Hi there,

I'm attempting to adopt `antsibull-nox` in the [`community.beszel`](https://github.com/ansible-collections/community.beszel/tree/0.6.0) collection and everything is working well apart from the `ansible-test-units` session doesn't seem to have a collection dependency which is required by the unit tests. The unit tests require the `community.internal_test_tools` collection which I have defined in [`tests/unit/requirements.yml`](https://github.com/ansible-collections/community.beszel/blob/0.6.0/tests/unit/requirements.yml) and can see this [should](https://github.com/ansible-community/antsibull-nox/blob/main/src/antsibull_nox/sessions/ansible_test.py#L367) be picked up. From the logs it is being installed into `.nox/.cache/downloaded` but it doesn't seem to be available inside the `ansible-test` container.

## Logs

```
community.beszel on  0.6.0 [?] is 📦 v0.6.0 via 🐍 v3.13.7 
❯ uv run nox -e ansible-test-units     
nox > Running session ansible-test-units-2.17-3.11
nox > Creating virtual environment (virtualenv) using python3.11 in .nox/ansible-test-units-2-17-3-11
nox > python -m pip install https://github.com/ansible/ansible/archive/stable-2.17.tar.gz -U
Downloading community.internal_test_tools to .nox/.cache/downloaded...
  Installing community.internal_test_tools via git+https://github.com/ansible-collections/community.internal_test_tools.git,main...
Installing community.internal_test_tools from .nox/.cache/downloaded/community-internal_test_tools-F3TXRKBDH2IF5KY5-0.18.4.tar.gz to .nox/.cache/extracted/2.17/community.internal_test_tools...
nox > cd /var/folders/04/bz4g6f_93ql7s4zv0jbz_q4h0000gn/T/antsibull-nox-ansible-test-units-2.17-3.11-collection-root-d_e2_kz4/ansible_collections/community/beszel
nox > ansible-test units --color -v --docker
Configured locale: en_US.UTF-8
Run command: docker -v
Detected "docker" container runtime version: Docker version 28.5.2, build ecc6942
Run command: docker info --format '{{ json . }}'
Run command: docker version --format '{{ json . }}'
Container runtime: docker client=28.5.2 server=28.5.2 cgroup=v2
Assuming Docker is available on localhost.
Run command: docker image inspect quay.io/ansible/default-test-container:9.6.0
Run command: docker image inspect quay.io/ansible/ansible-test-utility-container:2.0.0
Run command: docker run --volume /sys/fs/cgroup:/probe:ro --name ansible-test-probe-FoGyCK5l --rm quay.io/ansible/ansible-test-utility-container:2.0.0 sh -c 'aud ...
Container host audit status: EBADF (-9)
Container host max open files: 1048576
Container loginuid: unavailable
Starting new "ansible-test-controller-FoGyCK5l" container.
Run command: docker run --tmpfs /tmp:exec --tmpfs /run:exec --tmpfs /run/lock --volume /var/run/docker.sock:/var/run/docker.sock --cgroupns private -dt --ulimit  ...
Adding "ansible-test-controller-FoGyCK5l" to container database.
Run command: docker container inspect ansible-test-controller-FoGyCK5l
Run command: docker run --pid host --privileged --name ansible-test-init-controller-FoGyCK5l --rm quay.io/ansible/ansible-test-utility-container:2.0.0 nsenter -t ...
Stream command with data: docker exec -i ansible-test-controller-FoGyCK5l /bin/sh
Bootstrapping Python 3.12 at: /usr/bin/python3.12
Scanning collection root: /private/var/folders/04/bz4g6f_93ql7s4zv0jbz_q4h0000gn/T/antsibull-nox-ansible-test-units-2.17-3.11-collection-root-d_e2_kz4/ansible_collections
Including collection: community.beszel (88 files)
Creating a payload archive containing 1014 files...
Created a 1486334 byte payload archive containing 1014 files in 0 seconds.
Run command with stdin: docker exec -i ansible-test-controller-FoGyCK5l tar oxzf - -C /root
Creating container database.
Run command: docker exec ansible-test-controller-FoGyCK5l mkdir -p /root/ansible_collections/community/beszel/tests/output/junit /root/ansible_collections/commun ...
Run command: docker exec ansible-test-controller-FoGyCK5l chmod 777 /root/ansible_collections/community/beszel/tests/output/junit /root/ansible_collections/commu ...
Run command: docker exec ansible-test-controller-FoGyCK5l chmod 755 /root
Run command: docker exec ansible-test-controller-FoGyCK5l useradd pytest --create-home
Stream command: docker exec ansible-test-controller-FoGyCK5l /usr/bin/env ANSIBLE_TEST_CONTENT_ROOT=/root/ansible_collections/community/beszel LC_ALL=en_US.UTF-8 ...
Configured locale: en_US.UTF-8
Parsing container database.
Installing requirements for Python 3.7
Stream command with data: /usr/bin/python3.7
Execute command: /usr/bin/python3.7 /tmp/ansible-test-_3d13xz3-pip.py install --disable-pip-version-check -r requirements/units.txt -c requirements/constraints.txt
Run command: /usr/bin/python3.7 /root/ansible/test/lib/ansible_test/_util/target/tools/yamlcheck.py
Installing requirements for Python 3.8
Stream command with data: /usr/bin/python3.8
Execute command: /usr/bin/python3.8 /tmp/ansible-test-kcel2qq4-pip.py install --disable-pip-version-check -r requirements/units.txt -c requirements/constraints.txt
Run command: /usr/bin/python3.8 /root/ansible/test/lib/ansible_test/_util/target/tools/yamlcheck.py
Installing requirements for Python 3.9
Stream command with data: /usr/bin/python3.9
Execute command: /usr/bin/python3.9 /tmp/ansible-test-mmhtpi5i-pip.py install --disable-pip-version-check -r requirements/units.txt -c requirements/constraints.txt
Run command: /usr/bin/python3.9 /root/ansible/test/lib/ansible_test/_util/target/tools/yamlcheck.py
Installing requirements for Python 3.10
Stream command with data: /usr/bin/python3.10
Execute command: /usr/bin/python3.10 /tmp/ansible-test-c973rurk-pip.py install --disable-pip-version-check -r requirements/units.txt -c requirements/constraints.txt
Run command: /usr/bin/python3.10 /root/ansible/test/lib/ansible_test/_util/target/tools/yamlcheck.py
Installing requirements for Python 3.11
Stream command with data: /usr/bin/python3.11
Execute command: /usr/bin/python3.11 /tmp/ansible-test-oq2l1ufn-pip.py install --disable-pip-version-check -r requirements/units.txt -c requirements/constraints.txt
Run command: /usr/bin/python3.11 /root/ansible/test/lib/ansible_test/_util/target/tools/yamlcheck.py
Installing requirements for Python 3.12
Stream command with data: /usr/bin/python3.12
Execute command: /usr/bin/python3.12 /tmp/ansible-test-2lo_1csv-pip.py install --disable-pip-version-check -r requirements/units.txt -c requirements/constraints.txt
Run command: /usr/bin/python3.12 /root/ansible/test/lib/ansible_test/_util/target/tools/yamlcheck.py
Run command: docker container inspect ansible-test-controller-FoGyCK5l
Run command: docker network disconnect bridge ansible-test-controller-FoGyCK5l
Stream command: docker exec --user pytest ansible-test-controller-FoGyCK5l /usr/bin/env ANSIBLE_TEST_CONTENT_ROOT=/root/ansible_collections/community/beszel LC_A ...
Configured locale: en_US.UTF-8
Unit test modules with Python 3.7
Initializing "/tmp/ansible-test-zl32fms0-injector" as the temporary injector directory.
Injecting "/tmp/python-dacy02dk-ansible/python" as a execv wrapper for the "/usr/bin/python3.7" interpreter.
Stream command: pytest -r a -n auto --color yes -p no:cacheprovider -c /root/ansible/test/lib/ansible_test/_data/pytest/config/legacy.ini --junit-xml /root/ansib ...
============================= test session starts ==============================
platform linux -- Python 3.7.17, pytest-7.4.4, pluggy-1.2.0 -- /usr/bin/python3.7
rootdir: /root/ansible_collections/community/beszel
configfile: ../../../ansible/test/lib/ansible_test/_data/pytest/config/legacy.ini
plugins: mock-3.11.1, xdist-3.5.0
created: 10/10 workers
10 workers [0 items]

scheduling tests via LoadScheduling

==================================== ERRORS ====================================
__________ ERROR collecting tests/unit/plugins/modules/test_system.py __________
ImportError while importing test module '/root/ansible_collections/community/beszel/tests/unit/plugins/modules/test_system.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/usr/lib/python3.7/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
/root/ansible/test/lib/ansible_test/_util/target/pytest/plugins/ansible_pytest_collections.py:65: in exec_module
    assertion_rewriting_hook.exec_module(module)
tests/unit/plugins/modules/test_system.py:1: in <module>
    from ansible_collections.community.internal_test_tools.tests.unit.plugins.modules.utils import (
E   ModuleNotFoundError: No module named 'ansible_collections.community.internal_test_tools'
_______ ERROR collecting tests/unit/plugins/modules/test_system_info.py ________
ImportError while importing test module '/root/ansible_collections/community/beszel/tests/unit/plugins/modules/test_system_info.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/usr/lib/python3.7/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
/root/ansible/test/lib/ansible_test/_util/target/pytest/plugins/ansible_pytest_collections.py:65: in exec_module
    assertion_rewriting_hook.exec_module(module)
tests/unit/plugins/modules/test_system_info.py:1: in <module>
    from ansible_collections.community.internal_test_tools.tests.unit.plugins.modules.utils import (
E   ModuleNotFoundError: No module named 'ansible_collections.community.internal_test_tools'
_____ ERROR collecting tests/unit/plugins/modules/test_universal_token.py ______
ImportError while importing test module '/root/ansible_collections/community/beszel/tests/unit/plugins/modules/test_universal_token.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/usr/lib/python3.7/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
/root/ansible/test/lib/ansible_test/_util/target/pytest/plugins/ansible_pytest_collections.py:65: in exec_module
    assertion_rewriting_hook.exec_module(module)
tests/unit/plugins/modules/test_universal_token.py:1: in <module>
    from ansible_collections.community.internal_test_tools.tests.unit.plugins.modules.utils import (
E   ModuleNotFoundError: No module named 'ansible_collections.community.internal_test_tools'
- generated xml file: /root/ansible_collections/community/beszel/tests/output/junit/python3.7-modules-units.xml -
=========================== short test summary info ============================
ERROR tests/unit/plugins/modules/test_system.py
ERROR tests/unit/plugins/modules/test_system_info.py
ERROR tests/unit/plugins/modules/test_universal_token.py
============================== 3 errors in 0.38s ===============================
FATAL: Command "pytest -r a -n auto --color yes -p no:cacheprovider -c /root/ansible/test/lib/ansible_test/_data/pytest/config/legacy.ini --junit-xml /root/ansible_collections/community/beszel/tests/output/junit/python3.7-modules-units.xml --strict-markers --rootdir /root/ansible_collections/community/beszel --confcutdir /root/ansible_collections/community/beszel -v tests/unit/plugins/modules/test_system.py tests/unit/plugins/modules/test_system_info.py tests/unit/plugins/modules/test_universal_token.py" returned exit status 1.
```

## Version Details

```
community.beszel on  0.6.0 [!?] is 📦 v0.6.0 via 🐍 v3.13.7 took 8s 
❯ uv run nox --version                 
2025.11.12
ommunity.beszel on  0.6.0 [?] is 📦 v0.6.0 via 🐍 v3.13.7 
❯ uv pip list | grep antsibull      
antsibull-nox             1.3.2
```

--- a/changelogs/fragments/153-abs.yml
+++ b/changelogs/fragments/153-abs.yml
@@ -1,0 +1,3 @@
+bugfixes:
+  - "Avoid using relative symlinks to link from temporary collection root to collections. This can cause problems with non-canonical paths
+     (https://github.com/ansible-community/antsibull-nox/issues/152, https://github.com/ansible-community/antsibull-nox/pull/153)."

--- a/src/antsibull_nox/collection/install.py
+++ b/src/antsibull_nox/collection/install.py
@@ -445,9 +445,9 @@ def _add_all_dependencies(
 
 
 def _install_collection(collection: CollectionData, path: Path) -> None:
-    # Compute relative path
-    sym_path = collection.path.absolute().relative_to(path.parents[0], walk_up=True)
-    # Ensure that path is symlink with this relative path
+    # Compute absolute path
+    sym_path = collection.path.absolute()
+    # Ensure that path is symlink with this absolute path
     if path.is_symlink():
         if path.readlink() == sym_path:
             return
@@ -457,7 +457,9 @@ def _install_collection(collection: CollectionData, path: Path) -> None:
     path.symlink_to(sym_path)
 
 
-def _install_current_collection(collection: CollectionData, path: Path) -> None:
+def _install_current_collection(
+    collection: CollectionData, path: Path, *, use_relative_symlinks: bool
+) -> None:
     if path.exists() and (path.is_symlink() or not path.is_dir()):
         path.unlink()
     path.mkdir(exist_ok=True)
@@ -467,8 +469,12 @@ def _install_current_collection(collection: CollectionData, path: Path) -> None:
             continue
         dest_entry = path / source_entry.name
         # Compute relative path
-        sym_path = source_entry.relative_to(path, walk_up=True)
-        # Ensure that dest_entry is symlink with this relative path
+        sym_path = (
+            source_entry.relative_to(path, walk_up=True)
+            if use_relative_symlinks
+            else source_entry
+        )
+        # Ensure that dest_entry is symlink with this relative/absolute path
         if source_entry.name in present:
             present.remove(source_entry.name)
             if dest_entry.is_symlink() and dest_entry.readlink() == sym_path:
@@ -481,7 +487,11 @@ def _install_current_collection(collection: CollectionData, path: Path) -> None:
 
 
 def _install_collections(
-    collections: Iterable[CollectionData], root: Path, *, with_current: bool
+    collections: Iterable[CollectionData],
+    root: Path,
+    *,
+    with_current: bool,
+    use_relative_symlinks_for_current: bool = True,
 ) -> None:
     for collection in collections:
         namespace_dir = root / collection.namespace
@@ -490,7 +500,11 @@ def _install_collections(
         if not collection.current:
             _install_collection(collection, path)
         elif with_current:
-            _install_current_collection(collection, path)
+            _install_current_collection(
+                collection,
+                path,
+                use_relative_symlinks=use_relative_symlinks_for_current,
+            )
 
 
 def _extract_collections_from_extra_deps_file(path: str | os.PathLike) -> list[str]:
@@ -529,6 +543,7 @@ def setup_collections(
     extra_deps_files: list[str | os.PathLike] | None = None,
     global_cache_dir: Path,
     with_current: bool = True,
+    use_relative_symlinks_for_current: bool = True,
 ) -> SetupResult:
     """
     Setup all collections in a tree structure inside the destination directory.
@@ -572,7 +587,10 @@ def setup_collections(
             missing.remove(collection_data.full_name)
         missing.raise_error()
     _install_collections(
-        collections_to_install.values(), destination_root, with_current=with_current
+        collections_to_install.values(),
+        destination_root,
+        with_current=with_current,
+        use_relative_symlinks_for_current=use_relative_symlinks_for_current,
     )
     return SetupResult(
         root=destination_root,

--- a/src/antsibull_nox/paths.py
+++ b/src/antsibull_nox/paths.py
@@ -14,6 +14,7 @@ import atexit
 import functools
 import os
 import shutil
+from collections.abc import Sequence
 from pathlib import Path
 
 from antsibull_fileutils.copier import Copier, GitCopier
@@ -185,24 +186,27 @@ def copy_collection(
     copier.copy(source, destination, exclude_root=[".nox", ".tox"])
 
 
-def _is_acceptable_tempdir(directory: Path, exclude: Path) -> bool:
+def _is_acceptable_tempdir(directory: Path, excludes: Sequence[Path]) -> bool:
     # Avoid being inside an ansible_collections tree
     if not is_acceptable_tempdir(directory):
         return False
-    # Avoid something that's inside exclude
-    if directory.is_relative_to(exclude):
-        return False
+    # Avoid something that's inside one of the excludes
+    for exclude in excludes:
+        if directory.is_relative_to(exclude):
+            return False
     return True
 
 
-def get_outside_temp_directory(exclude: Path | None = None) -> Path:
+def get_outside_temp_directory(exclude: Sequence[Path] | Path | None = None) -> Path:
     """
     Find a temporary directory root that's outside the provided path.
 
     Raises ``ValueError`` in case no candidate was found.
     """
     if exclude is None:
-        exclude = Path.cwd()
+        exclude = [Path.cwd()]
+    elif isinstance(exclude, Path):
+        exclude = [exclude]
     return find_tempdir(lambda path: _is_acceptable_tempdir(path, exclude))
 
 

--- a/src/antsibull_nox/sessions/ee_check.py
+++ b/src/antsibull_nox/sessions/ee_check.py
@@ -199,7 +199,9 @@ def add_execution_environment_session(
         )
 
         playbook_dir = Path.cwd()
-        temp_dir = get_outside_temp_directory(playbook_dir.absolute())
+        temp_dir = get_outside_temp_directory(
+            [playbook_dir.absolute(), playbook_dir.resolve()]
+        )
 
         for playbook in execution_environment.test_playbooks:
             env = {"TMPDIR": str(temp_dir)}

